docs/refman/ppr-passwd.1.pod
#
# mouse:~ppr/src/docs/refman/ppr-passwd.8.pod
# Copyright 1995--2003, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 9 March 2003.
#

=head1 NAME

B<ppr-passwd> - Change a password for the PPR web interface

=head1 SYNOPSIS

B<ppr-passwd> [B<--add>] [B<--delete>] I<username>

=head1 DESCRIPTION

This program is used to edit the password file for the PPR web interface. 
The web interface needs its own password file in order to support MD5 Digest
HTTP authentication.

If no I<username> is specified, then the invoking users own password will be
changed.  Users listed in B</etc/ppr/acl/passwd.allow> (as well as root and
ppr) may change other people's passwords and needn't know the old password.

This program supports the following options:

=over 4

=item B<--add>

=item B<-a>

Add the user to the password file if he doesn't already exist.

=item B<--delete>

=item B<-x>

Delete the user from the password database.

=item B<--help>

Prints a short usage summary.

=item B<--version>

Prints the PPR version number and copyright information.

=back

=head1 FILES

The password file whic this program modifies is F</etc/ppr/htpasswd>.

=head1 SEE ALSO

"PPR, a PostScript Print Spooler"

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

www/ppr-passwd.c
/*
** mouse:~ppr/src/www/ppr-passwd.c
** Copyright 1995--2003, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
**
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 10 September 2003.
*/

/*
** This program updates /etc/ppr/htpasswd, a file which contains MD5 digests
** for use with MD5 Digest HTTP authentication.
*/

#include "before_system.h"
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <pwd.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "gu_md5.h"
#include "global_defines.h"
#include "util_exits.h"
#include "version.h"

const char myname[] = "ppr-passwd";
const char htpasswd[] = CONFDIR"/htpasswd";
const char htpasswd_new[] = CONFDIR"/htpasswd_new";
const char realm[] = "printing";

static int old_password(const char username[], const char realm[], const char digested[]);
static int new_password(FILE *wfile, const char username[], const char realm[]);
static void md5_digest(char *buffer, const char username[], const char realm[], const char password[]);

/*
** Command line options:
*/
static const char *option_chars = "ax";
static const struct gu_getopt_opt option_words[] =
		{
		{"add", 'a', FALSE},
		{"delete", 'x', FALSE},
		{"help", 9000, FALSE},
		{"version", 9001, FALSE},
		{(char*)NULL, 0, FALSE}
		} ;

/*
** Print help.
*/
static void help_usage(FILE *outfile)
	{
	fputs(_("Usage: ppr-passwd <username>\n"), outfile);

	fputc('\n', outfile);

	fputs(_("Valid switches:\n"), outfile);

	fputs(_(	"\t--version\n"
				"\t--help\n"), outfile);
	}

int main(int argc, char *argv[])
	{
	gu_boolean privileged;
	const char *invoking_user;
	const char *username;
	gu_boolean opt_add = FALSE;
	gu_boolean opt_delete = FALSE;
	int rfd, wfd;
	FILE *rfile, *wfile;

	/* Initialize international messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_MESSAGES, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	{
	struct passwd *pw;
	uid_t uid = getuid();
	if(!(pw = getpwuid(uid)))
		{
		fprintf(stderr, _("%s: getpwuid(%ld) failed, errno=%d (%s)\n"), myname, (long)getuid(), errno, gu_strerror(errno));
		return EXIT_NOTFOUND;
		}
	invoking_user = gu_strdup(pw->pw_name);

	/* Check if this user is root or ppr or is listed in /etc/ppr/acl/passwd.allow. */
	privileged = user_acl_allows(invoking_user, "passwd");

	/* Prevent user from sending us signals. */
	if(setreuid(geteuid(), -1) == -1)
		{
		fprintf(stderr, "%s: setreuid(%ld, %ld) failed, errno=%d (%s)\n", myname, (long)geteuid(), (long)-1, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}
	}

	/* Parse the options. */
	{
	struct gu_getopt_state getopt_state;
	int optchar;
	gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);
	while((optchar = ppr_getopt(&getopt_state)) != -1)
		{
		switch(optchar)
			{
			case 'a':					/* --add */
				opt_add = TRUE;
				break;
			
			case 'x':					/* --delete */
				opt_delete = TRUE;
				break;

			case 9000:					/* --help */
				help_usage(stdout);
				exit(EXIT_OK);

			case 9001:					/* --version */
				puts(VERSION);
				puts(COPYRIGHT);
				puts(AUTHOR);
				exit(EXIT_OK);

			default:					/* other getopt errors or missing case */
				gu_getopt_default(myname, optchar, &getopt_state, stderr);
				exit(EXIT_SYNTAX);
				break;
			}
		}

	if(argc > getopt_state.optind)
		{
		if(!privileged)
			{
			fprintf(stderr, _("%s: only privileged users may change passwords for others.\n"), myname);
			return EXIT_DENIED;
			}
		username = argv[getopt_state.optind];
		}
	else
		{
		username = invoking_user;
		}
	}

    /* Catching this undefined combination is good policy. */
	if(opt_add && opt_delete)
		{
		fprintf(stderr, _("%s: cannot add and delete a user at the same time.\n"), myname);
		return EXIT_ERROR;
		}

	/*
		Ordinary users aren't allowed to add entries, even for themselves 
		because then someone stepping up to an unattended terminal logged
		into the account of a person who did not use the PPR web interface
		regularly could open that account to web printer management, setting
		a password known to the interloper.
		
		Note that non-privileged users must know their passwords in order
		to change them.
	*/
	if((opt_add || opt_delete) && !privileged)
		{
		fprintf(stderr, _("%s: only privileged users may add and delete entries.\n"), myname);
		return EXIT_DENIED;
		}

	if((rfd = open(htpasswd, O_RDWR)) == -1)
		{
		fprintf(stderr, _("%s: can't open \"%s\", errno=%d (%s).\n"), myname, htpasswd, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}

	if(gu_lock_exclusive(rfd, TRUE) == -1)
		{
		fprintf(stderr, _("%s: failed to get exclusive lock on \"%s\", errno=%d (%s).\n"), myname, htpasswd, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}

	if((wfd = open(htpasswd_new, (O_WRONLY | O_CREAT | O_EXCL), UNIX_640)) == -1)
		{
		fprintf(stderr, _("%s: can't create \"%s\", errno=%d (%s)\n"), myname, htpasswd_new, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}

	if(!(rfile = fdopen(rfd, "r")) || !(wfile = fdopen(wfd, "w")))
		{
		fprintf(stderr, _("%s: fdopen() failed, errno=%d (%s)\n"), myname, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}

	{
	int ret = -1;
	{
	char *line = NULL;
	int line_available = 80;
	gu_boolean found = FALSE;
	char *p, *f1, *f2, *f3;
	int linenum = 0;

	/* Step thru all the lines of /etc/ppr/htpasswd. */
	while((p = line = gu_getline(line, &line_available, rfile)))
		{
		linenum++;

		/* Locate the three fields.  Each line looks like this:
				david:printing:c2e93b4cf3b62bdfd5b4680c2a1b8460
		*/
		if(!(f1 = gu_strsep(&p,":")) || !(f2 = gu_strsep(&p,":")) || !(f3 = gu_strsep(&p, ":")))
			{
			fprintf(stderr, _("%s: line %d of %s is invalid, aborting\n"), myname, linenum, htpasswd);
			break;
			}

		/* If this is the one we are looking for, */
		if(strcmp(f1, username) == 0 && strcmp(f2, realm) == 0)
			{
			found = TRUE;

			if(!opt_delete)
				{
				/* Non-privileged users must prove that they know the old password. */
				if(!privileged)
					{
					if((ret = old_password(username, realm, f3)) != 0)
						break;
					}
				
				/* Get new password from user and write out the user:realm:digest line. */
				if((ret = new_password(wfile, username, realm)) != 0)
					break;
				}
			}

		/* Others just get copied. */
		else
			{
			if(fprintf(wfile, "%s:%s:%s\n", f1, f2, f3) < 0)
				{
				fprintf(stderr, "%s: fprintf() failed\n", myname);
				break;
				}
			}
		}

	/* Line won't have been freed if break in loop. */
	if(line)
		gu_free(line);

	/* If we haven't already changed an existing password, */
	if(!found)
		{
		if(!opt_add)
			{
			fprintf(stderr, _("%s: user \"%s\" not found in \"%s\".\n"), myname, username, htpasswd);
			ret = EXIT_NOTFOUND;
			}
		else
			{
			ret = new_password(wfile, username, realm);
			}
		}

	}

	/* /etc/ppr/htpasswd_new */
	if(fclose(wfile))
		{
		fprintf(stderr, _("%s: fclose() failed, errno=%d (%s)\n"), myname, errno, gu_strerror(errno));
		ret = -1;
		}

	/* /etc/ppr/htpasswd */
	if(fclose(rfile))
		{
		fprintf(stderr, _("%s: fclose() failed, errno=%d (%s)\n"), myname, errno, gu_strerror(errno));
		ret = -1;
		}

	/* If we failed, get rid of /etc/ppr/htpasswd_new. */
	if(ret != 0)
		{
		if(unlink(htpasswd_new) == -1)
			{
			fprintf(stderr, _("%s: unlink(\"%s\") failed, errno=%d (%s)\n"), myname, htpasswd_new, errno, gu_strerror(errno));
			}
		printf(_("No changes made.\n"));
		}

	/* Otherwise rename /etc/ppr/htpasswd_new --> /etc/ppr/htpasswd (an atomic operation). */
	else
		{
		if(rename(htpasswd_new, htpasswd) == -1)
			{
			fprintf(stderr, _("%s: rename(\"%s\", \"%s\") failed, errno=%d (%s)\n"), myname, htpasswd_new, htpasswd, errno, gu_strerror(errno));
			ret = -1;
			}
		printf(_("Password file updated sucessfully.\n"));
		}

	if(ret > 0)
		return ret;
	if(ret == -1)
		return EXIT_INTERNAL;
	}

	return EXIT_OK;
	} /* end of main() */

static int old_password(const char username[], const char realm[], const char digested[])
	{
	char prompt[80];
	char *pass_ptr = NULL;
	char old_digested[33];

	snprintf(prompt, sizeof(prompt), _("Please enter old password for user %s in realm %s: "), username, realm);

	while(!pass_ptr)
		{
		if(!(pass_ptr = getpass(prompt)))
			return -1;
		if(strlen(pass_ptr) == 0)
			return EXIT_USER_ABORT;
		}

	md5_digest(old_digested, username, realm, pass_ptr);

	if(strcmp(old_digested, digested) != 0)
		{
		printf(_("Password incorrect, access denied.\n"));
		return EXIT_DENIED;
		}

	return 0;
	}

static int new_password(FILE *wfile, const char username[], const char realm[])
	{
	char prompt[80];
	char *pass_ptr_first = NULL;
	char *pass_ptr = NULL;
	char crypted[33];
	gu_boolean match;

	snprintf(prompt, sizeof(prompt), _("Please enter new password for user %s in realm %s: "), username, realm);

	do	{
		while(!pass_ptr_first)
			{
			if(!(pass_ptr_first = getpass(prompt)))
				return -1;
			if(strlen(pass_ptr_first) == 0)
				return EXIT_USER_ABORT;
			if(strlen(pass_ptr_first) < 6)
				{
				printf(_("That password is too short.  Please pick another.\n"));
				pass_ptr = NULL;
				continue;
				}
			break;
			}

		pass_ptr_first = gu_strdup(pass_ptr_first);

		while(!pass_ptr)
			{
			if(!(pass_ptr = getpass(_("Please verify new password: "))))
				break;
			}

		match = (pass_ptr && strcmp(pass_ptr_first, pass_ptr) == 0);
		memset(pass_ptr_first, 0, strlen(pass_ptr_first));
		gu_free(pass_ptr_first);
		pass_ptr_first = NULL;

		if(!match)
			 {
			 fputc('\n', stdout);
			 printf(_("Passwords do not match, please try again.\n"));
			 pass_ptr = NULL;
			 }
		} while(!match);

	/* Create the MD5 signiture and then zero the cleartext password out
	   to keep it from prying eyes.
	   */
	md5_digest(crypted, username, realm, pass_ptr);
	memset(pass_ptr, 0, strlen(pass_ptr));

	if(fprintf(wfile, "%s:%s:%s\n", username, realm, crypted) < 0)
		{
		fprintf(stderr, "%s: fprintf() failed\n", myname);
		return -1;
		}

	return 0;
	}

static void md5_digest(char *buffer, const char username[], const char realm[], const char password[])
	{
	md5_state_t state;
	md5_byte_t digest[16];
	int i;

	md5_init(&state);
	md5_append(&state, (const md5_byte_t *)username, strlen(username));
	md5_append(&state, (const md5_byte_t *)":", strlen(":"));
	md5_append(&state, (const md5_byte_t *)realm, strlen(realm));
	md5_append(&state, (const md5_byte_t *)":", strlen(":"));
	md5_append(&state, (const md5_byte_t *)password, strlen(password));
	md5_finish(&state, digest);

	for(i = 0; i < 16; i++)
		snprintf(&buffer[i * 2], 3, "%02x", digest[i]);
	}

/* end of file */
